Cubical/Foundations/Isomorphism.agda

```diff
@@ -0,0 +1,121 @@
+  isPropIsContr (u0 .equiv-proof y) (u1 .equiv-proof y) i
+
+-- Direct proof that computes quite ok (can be optimized further if
+-- necessary, see:
+-- https://github.com/mortberg/cubicaltt/blob/pi4s3_dimclosures/examples/brunerie2.ctt#L562
+isPropIsEquiv : ∀ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} (f : A → B) → isProp (isEquiv f)
+equiv-proof (isPropIsEquiv f p q i) y =
+  let p2 = p .equiv-proof y .snd
+      q2 = q .equiv-proof y .snd
+  in p2 (q .equiv-proof y .fst) i
+   , λ w j → hcomp (λ k → λ { (i = i0) → p2 w j
+                            ; (i = i1) → q2 w (j ∨ ~ k)
+                            ; (j = i0) → p2 (q2 w (~ k)) i
+                            ; (j = i1) → w })
+                   (p2 w (i ∨ j))
+
+equivEq : ∀ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} (e f : A ≃ B) → (h : e .fst ≡ f .fst) → e ≡ f
+equivEq e f h = λ i → (h i) , isProp→PathP isPropIsEquiv h (e .snd) (f .snd) i
+
+
+-- Section and retract
+module _ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} where
+  section : (f : A → B) → (g : B → A) → Set ℓ'
+  section f g = ∀ b → f (g b) ≡ b
+
+  -- NB: `g` is the retraction!
+  retract : (f : A → B) → (g : B → A) → Set ℓ
+  retract f g = ∀ a → g (f a) ≡ a
+
+-- Any iso is an equivalence
+module _ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} (f : A → B) (g : B → A)
+         (s : section f g) (t : retract f g) where
+
+  private
+    module _ (y : B) (x0 x1 : A) (p0 : f x0 ≡ y) (p1 : f x1 ≡ y) where
+      fill0 : I → I → A
+      fill0 i = hfill (λ k → λ { (i = i1) → t x0 k
+                               ; (i = i0) → g y })
+                      (inc (g (p0 (~ i))))
+
+      fill1 : I → I → A
+      fill1 i = hfill (λ k → λ { (i = i1) → t x1 k
+                               ; (i = i0) → g y })
+                      (inc (g (p1 (~ i))))
+
+      fill2 : I → I → A
+      fill2 i = hfill (λ k → λ { (i = i1) → fill1 k i1
+                               ; (i = i0) → fill0 k i1 })
+                      (inc (g y))
+
+      p : x0 ≡ x1
+      p i = fill2 i i1
+
+      sq : I → I → A
+      sq i j = hcomp (λ k → λ { (i = i1) → fill1 j (~ k)
+                              ; (i = i0) → fill0 j (~ k)
+                              ; (j = i1) → t (fill2 i i1) (~ k)
+                              ; (j = i0) → g y })
+                     (fill2 i j)
+
+      sq1 : I → I → B
+      sq1 i j = hcomp (λ k → λ { (i = i1) → s (p1 (~ j)) k
+                               ; (i = i0) → s (p0 (~ j)) k
+                               ; (j = i1) → s (f (p i)) k
+                               ; (j = i0) → s y k })
+                      (f (sq i j))
+
+      lemIso : (x0 , p0) ≡ (x1 , p1)
+      lemIso i .fst = p i
+      lemIso i .snd = λ j → sq1 i (~ j)
+
+  isoToIsEquiv : isEquiv f
+  isoToIsEquiv .equiv-proof y .fst .fst = g y
+  isoToIsEquiv .equiv-proof y .fst .snd = s y
+  isoToIsEquiv .equiv-proof y .snd z = lemIso y (g y) (fst z) (s y) (snd z)
+
+  isoToEquiv : A ≃ B
+  isoToEquiv = _ , isoToIsEquiv
+
+module _ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} (w : A ≃ B) where
```

Comment:
Let's move this back to Equiv as well. The secEq and retEq statements will have to be unfolded, but that's fine